src/main/scala/beam/physsim/PickUpDropOffCollector.scala

```diff
@@ -0,0 +1,270 @@
+import beam.agentsim.events.PathTraversalEvent
+import beam.router.Modes
+import beam.sim.config.BeamConfig
+import com.typesafe.scalalogging.LazyLogging
+import org.matsim.api.core.v01.Id
+import org.matsim.api.core.v01.events.{Event, PersonEntersVehicleEvent, PersonLeavesVehicleEvent}
+import org.matsim.core.controler.events.{IterationEndsEvent, IterationStartsEvent}
+import org.matsim.core.controler.listener.{IterationEndsListener, IterationStartsListener}
+import org.matsim.core.events.handler.BasicEventHandler
+import org.matsim.vehicles.Vehicle
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+class PickUpDropOffCollector(vehicleTypes: Map[Id[BeamVehicleType], BeamVehicleType])
+    extends BasicEventHandler
+    with IterationStartsListener
+    with LazyLogging {
+
+  private val vehicleToPickUpsDropOffs: mutable.HashMap[Id[Vehicle], VehiclePickUpsDropOffs] = mutable.HashMap.empty
+  private val personsPickUps: mutable.ListBuffer[PersonPickUp] = mutable.ListBuffer.empty
+  private val personsDropOffs: mutable.ListBuffer[PersonDropOff] = mutable.ListBuffer.empty
+
+  private val vehiclesToIgnore: mutable.HashSet[Id[Vehicle]] = mutable.HashSet.empty
+  private val vehicleTypeToCAV: mutable.HashMap[String, Boolean] = mutable.HashMap.empty
+
+  private def isRH(vehicleId: Id[Vehicle]): Boolean = vehicleId.toString.startsWith("rideHailVehicle")
+
+  private var maybePickUpDropOffHolder: Option[PickUpDropOffHolder] = None
+
+  private def isCAV(vehicleType: String): Boolean = {
+    vehicleTypeToCAV.get(vehicleType) match {
+      case None =>
+        val beamVehicleTypeId = Id.create(vehicleType, classOf[BeamVehicleType])
+        val isCaccEnabled = vehicleTypes.get(beamVehicleTypeId) match {
+          case Some(beamVehicleType) => beamVehicleType.isCaccEnabled
+          case _                     => false
+        }
+
+        vehicleTypeToCAV.put(vehicleType, isCaccEnabled)
+        isCaccEnabled
+
+      case Some(isCaccEnabled) => isCaccEnabled
+    }
+  }
+
+  private def pteIsSelected(pte: PathTraversalEvent): Boolean = {
+    val mode = pte.mode
+    mode match {
+      case Modes.BeamMode.CAR if isRH(pte.vehicleId) || isCAV(pte.vehicleType) => true
+      case Modes.BeamMode.CAV                                                  => true
+      case Modes.BeamMode.RIDE_HAIL                                            => true
+      case Modes.BeamMode.RIDE_HAIL_POOLED                                     => true
+      case Modes.BeamMode.RIDE_HAIL_TRANSIT                                    => true
+      case _                                                                   => false
+    }
+  }
+
+  override def handleEvent(event: Event): Unit = {
+    def notIgnored(vehicleId: Id[Vehicle]): Boolean = {
+      val isIgnored = vehiclesToIgnore.contains(vehicleId)
+      !isIgnored
+    }
+
+    event match {
+      case pte: PathTraversalEvent if pteIsSelected(pte) && notIgnored(pte.vehicleId) =>
```

Comment:
As far as I can tell, ignored is really just where pte is NOT selected? If so, then this additional check only adds a jump that is not needed. If you switched the order then it might save a call to pteIsSelected if we feel it is less performant (debatable). I would argue the notIgnored can be dropped altogether in this condition.

Reply:
there are two separate functions, one is to check if an event is what we are looking for, another is to check if a vehicle should be ignored. 
I have renamed them to not be confusing anymore.

Reply:
The naming wasn't the issue, but instead the logic itself. It's more of an optimization on performance, so maybe readability is more important and I'll resolve  - feel free to change or not :)